File: site/script/loading_screen.rb
class LoadingScreen
  attr_reader :element

  def initialize
    @element = $window.document.create_element('div')
    @element.class_list.add('loading-screen', 'hidden', 'transition')
    @element.id = "loading_screen#{(Time.now.to_f * 1000).to_i}"

    container = $window.document.create_element('div')
    container.class_list.add('loading-container')
    @element.append_child(container)

    text = $window.document.create_element('p')
    text.inner_text = 'Loading data...'
    container.append_child(text)

    bar = $window.document.create_element('div')
    bar.class_list.add('loading-bar')
    container.append_child(bar)
  end

  def show
    @can_hide_at = Time.now + 0.25
    $window.set_timeout(0) { @element.class_list.remove('hidden') }
  end

  def hide
    return if @element.class_list.contains('hidden')

    time_left_until_hide = @can_hide_at - Time.now
    if time_left_until_hide.to_f > 0
      new_time_left = time_left_until_hide * 1000
      $window.set_timeout(new_time_left) { hide } 
      return
    end

    @element.class_list.add('hidden')
    $window.set_timeout(250) { @element.parent_node.remove_child(@element) }
  end

end # LoadingScreen

File: site/scripts/loading-screen.js
var LoadingScreen = {
  MINIMUM_SHOW_TIME: 1000,

  show: function() {
    if (!this._element) {
      this._initializeElement();
    }

    this._canHideAt = Date.now() + this.MINIMUM_SHOW_TIME;
    document.body.appendChild(this._element);
    window.setTimeout(function() {
      this._element.classList.remove('hidden');
    }.bind(this), 0);
  },

  hide: function() {
    if (!this._canHideAt || !this._element) {
      return;
    }

    var timeLeftUntilHide = this._canHideAt - Date.now();
    if (timeLeftUntilHide > 0) {
      window.setTimeout(function() {
        this.hide();
      }.bind(this), timeLeftUntilHide);

      return;
    }

    this._element.classList.add('hidden');
    window.setTimeout(function() {
      this._element.parentNode.removeChild(this._element);
    }.bind(this), 400);
  },

  _initializeElement: function() {
    this._element = document.createElement('div');
    this._element.classList.add('loading-screen', 'hidden', 'transition');

    var container = document.createElement('div');
    container.classList.add('loading-container');
    this._element.appendChild(container);

    text = document.createElement('p');
    text.innerHTML = 'Loading data...';
    container.appendChild(text);

    bar = document.createElement('div');
    bar.classList.add('loading-bar');
    container.appendChild(bar);
  }
};